onnxruntime/core/graph/graph_viewer.cc

```diff
@@ -35,6 +35,17 @@ struct PriorityNodeCompare {
       return n1->Priority() > n2->Priority();
     }
 
+    // nodes of forward pass will be output first
+    auto n1_attrs = n1->GetAttributes();
+    auto n2_attrs = n2->GetAttributes();
+    int64_t n1_is_forward = static_cast<int64_t>(n1_attrs.find(kBackwardNodeAttributeName) == n1_attrs.cend()) ||
```

Comment:
I dont quite understand how this value is being calculated. What will be n1_is_forward when "(n1_attrs.find(kBackwardNodeAttributeName) == n1_attrs.cend())" is True and when it is False.

Reply:
You can check line 25 and 26. If the function return true, n2 will be output first, e.g. before n1 in the topo order. So in this case, if n2 is forward pass, but n1 is not, then n2 should be before n1.